### ISSUE TITLE
[P0] Simplify dataset structure

### ISSUE THREAD

**Issue by aryamanarora** [stanfordnlp/pyreft#43]
From an email today:
> On `pyreft/dataset.py`, me and Zen discussed today that it’s slightly convoluted and probably the most difficult part of the library to deal with. Basically, the additional complexity comes from:
> 1. `intervention_locations`: we need to compute the locations to perform interventions at, which currently is only applied to the prompt – the shape is [num_interventions, batch_size, num_locations]. When tying prefix/suffix weights, this means num_interventions = 2 * layers (i.e. prefix and suffix positions for each layers).  num_locations is the positions we intervene at for each set of intervention weights, so for first p or last s positions. Since we need this to be a fixed-size tensor, we add a pad token to the start of the sequence, add 1 to all locations in the tensor, and pad with 0s.
> 2. `subspaces`: how to partition the subspaces for multi-task training. We use this in our ReFT composition experiments but it’s not needed for what you’re doing (for now at least), so setting it to None is good.
> 
> We don’t want the user to have to think about this when adapting to new tasks, so I’m going to be making two base classes that compute this stuff automatically. One will be `ReftPromptlessDataset`, which will just take a single text input (no prompt template) and compute these. The other will be `ReftPromptDataset`, with a prompt and a completion, and we will compute the intervention only on the prompt. Hopefully you can easily inherit from one of these to make your dataset w/o worrying as much about interventions or subspaces.

**Comment by aryamanarora** [comment]
I restructured the datasets so that intervention locations, subspace stuff, and 1 token padding on the left are handled by `ReftDataset`. Every dataset class should just inherit this class and define the abstract method `tokenize()` which is how the dataset is processed (tokenization, attention masks, prompt templates) before the computation of all the stuff I mentioned above.

What remains to be done?

- [ ] Implement preference and reward dataset classes.
- [ ] Fix LoReFT example code to use these classes instead of super-specific ones (e.g. the GLUE stuff).

**Comment by aryamanarora** [comment]
I think we fixed this gg
